homeassistant/components/withings/sensor.py

```diff
@@ -0,0 +1,639 @@
+        """Update the data."""
+        _LOGGER.debug(
+            "Async update slug: %s, measurement: %s, user_id: %s",
+            self._slug, self._attribute.measurement, self._user_id
+        )
+
+        if isinstance(self._attribute, WithingsMeasureAttribute):
+            _LOGGER.debug("Updating measures state")
+            self._data_manager.update_measures()
+            await self.async_update_measure(self._data_manager.measures)
+
+        elif isinstance(self._attribute, WithingsSleepStateAttribute):
+            _LOGGER.debug("Updating sleep state")
+            self._data_manager.update_sleep()
+            await self.async_update_sleep_state(self._data_manager.sleep)
+
+        elif isinstance(self._attribute, WithingsSleepSummaryAttribute):
+            _LOGGER.debug("Updating sleep summary state")
+            self._data_manager.update_sleep_summary()
+            await self.async_update_sleep_summary(
+                self._data_manager.sleep_summary
+            )
+
+    async def async_update_measure(self, data) -> None:
+        """Update the measures data."""
+        if data is None:
+            _LOGGER.error(
+                "Provided data is None. Setting state to %s",
+                None
+            )
+            self._state = None
+            return
+
+        measurement = self._attribute.measurement
+        measure_type = self._attribute.measure_type
+        unit_of_measurement = self._attribute.unit_of_measurement
+
+        _LOGGER.debug(
+            "Finding the unambiguous measure group with measure_type: %s",
+            measure_type
+        )
+        measure_groups = [g for g in data if (
+            not g.is_ambiguous() and
+            g.get_measure(measure_type) is not None
+        )]
+
+        if not measure_groups:
+            _LOGGER.warning(
+                "No measure groups found, setting state to %s",
+                None
+            )
+            self._state = None
+            return
+
+        _LOGGER.debug(
+            "Sorting list of %s measure groups by date created (DESC)",
+            len(measure_groups)
+        )
+        measure_groups.sort(key=(lambda g: g.created), reverse=True)
+
+        value = measure_groups[0].get_measure(measure_type)
+
+        _LOGGER.debug(
+            "Determining state for measurement: %s, measure_type: %s, unit_of_measurement: %s, value: %s",  # pylint: disable=line-too-long  # noqa: E501
+            measurement, measure_type, unit_of_measurement, value
+        )
+
+        if unit_of_measurement is const.UOM_MASS_KG:
+            state = round(value, 1)
+
+        elif unit_of_measurement is const.UOM_MASS_LB:
+            state = round(value * 2.205, 2)
+
+        elif unit_of_measurement is const.UOM_MASS_STONE:
+            state = round(value * 0.157473, 2)
+
+        elif unit_of_measurement is const.UOM_LENGTH_M:
+            state = round(value, 2)
+
+        elif unit_of_measurement is const.UOM_LENGTH_CM:
```

Comment:
We should remove all the custom unit conversions that are done

Reply:
An integration should represent the data as-is from the data source. 

If conversion to other units is something that needs to be available for an entity, we should change that in a way that it's available for all entities of a specific device class, not handle it at a per integration level.